libtenzir/builtins/operators/import.cpp

```diff
@@ -76,45 +30,48 @@ class import_operator final : public crtp_operator<import_operator> {
     // TODO: Some of the the requests this operator makes are blocking, so we
     // have to create a scoped actor here; once the operator API uses async we
     // can offer a better mechanism here.
-    auto blocking_self = caf::scoped_actor{ctrl.self().system()};
-    auto components
-      = get_node_components<importer_actor>(blocking_self, ctrl.node());
-    if (!components) {
-      ctrl.abort(std::move(components.error()));
+    auto result = import_stream::make(ctrl.self(), ctrl.node());
+    if (not result) {
+      ctrl.abort(result.error());
       co_return;
     }
-    co_yield {};
-    auto [importer] = std::move(*components);
-    auto num_events = uint64_t{};
-    auto source = caf::detail::make_stream_source<import_source_driver>(
-      &ctrl.self(), input, num_events, ctrl);
-    source->add_outbound_path(importer);
-    for (const auto& plugin : plugins::get<analyzer_plugin>()) {
-      // We can safely assert that the analyzer was already initialized. The
-      // pipeline API guarantees that remote operators run after the node was
-      // successfully initialized, which implies that analyzers have been
-      // initialized as well.
-      auto analyzer = plugin->analyzer();
-      TENZIR_ASSERT(analyzer);
-      source->add_outbound_path(analyzer);
-    }
-    while (input.unsafe_current() != input.end()) {
-      if (source->generate_messages()) {
-        source->out().emit_batches();
+    auto stream = std::move(*result);
+    auto total_events = size_t{0};
+    for (auto&& slice : input) {
+      if (stream.has_ended()) {
+        ctrl.abort(caf::make_error(
+          ec::unspecified,
+          fmt::format("unexpected import stream end: {}", stream.error())));
+        co_return;
       }
+      if (slice.rows() == 0) {
+        co_yield {};
+        continue;
+      }
+      total_events += slice.rows();
+      stream.enqueue(std::move(slice));
+      while (stream.enqueued() > 0) {
+        co_yield {};
+      }
+    }
```

Comment:
I'm not sure if this is ideal. Previously, we had the following:

~~~
while (input.unsafe_current() != input.end()) {
  if (source->generate_messages()) {
    source->out().emit_batches();
  }
  co_yield {};
}
~~~

But IIRC, this was mostly to work around a bug. Still, this refactoring changes the logic around stalling and backpressure quite a bit, so this needs a second critical look.